design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,217 @@
+* Setting an annotation in kubevirt CR is unconventional. 
+* A cluster-wide configuration such as this may be CPU-wasteful for heterogeneous clusters, where node-level granularity is needed.
+
+# Fix Alternatives
+## Alternative #1 Adding `EmulatorThreadEvenParity` annotation per VMI
+Another way to do it is by adding the annotation on the VMI's annotations:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    metadata:
+      annotations:
+        kubevirt.io/EmulatorThreadEvenParity:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+Setting this will set the same behavior as on the suggested solution on virt-launcher and virt-handler.
+As opposed to the suggested solution where the user needs to add this annotation once on the kubevirt CR, it is possible to make this behavior VMI specific. This removes the assumption that the Nodes are homogeneous.
+The annotation will still be accompanied by the feature-gate configuration.
+
+Pros:
+* Matches the current crio convention of adding annotations in order to get high-performance pods, like `cpu-load-balancing.crio.io: disable`
+* Reduces the “wasted” CPUs to minimum possible per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs.
+* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.
+
+Cons:
+* We’re adding yet another semi alpha knob to an already complicated set of knobs.
+
+## Alternative #2 Configure both pod and Guest
+It is possible to use both the current Guest API and the old yet existing pod API, like so:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        resources:
+          requests:
+            cpu: '10'
+          limits:
+            cpu: '10'
+        cpu:
+          cores: 4
+          sockets: 1
+          threads: 2
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+```
+In order to do that the current validating webhooks regarding dedicatedCpuPlacement and IsolatedEmulatorThread will have to be loosened.
+
+Pros:
+* This option does not introduce any alpha knobs.
+
+Cons:
+* This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse. Loosening the validating webhook doesn't make sense in the general case.
+* KubeVirt will need to adjust internal assumptions that are based on the fact that VMs {req, limits}.cpu = vCPUs. Specifically, how the housekeeping cgroup is being formed, but possibly other aspects.
+* VMs that were defined like this will have to change when the long term solution will arrive (backwards-compatibility issue)
```

Comment:
```suggestion
* VMs that were defined like this will have to change when the long term solution will arrive (backwards-compatibility issue)
```
Why? Do you think that VMI.spec.resources.requests.cpu would ever be deprecated?

Reply:
TBH I don't remember why this was written so perhaps it's better removed.
DONE.